### PR TITLE
Rework canBeTruthy to not rely on dropSubtypesOf

### PR DESCRIPTION
**Pull request by elliottt**
### Motivation

Rework `canBeTruthy` to not rely on `dropSubtypesOf`, as it's potentially very expensive when it causes a large Enum to be expanded out.

### Test plan

See included automated tests.

### CODE REVIEW

File: infer/environment.cc [hunk 1]
```diff
@@ -599,14 +602,15 @@ void Environment::mergeWith(core::Context ctx, const Environment &other, core::L
             pair.second.knownTruthy = pair.second.knownTruthy && other.getKnownTruthy(var);
         } else {
             thisTO = otherTO;
+            sameType = false;
```
Comment (DarkDimius) [inline infer/environment.cc]: I think this can be check if `otherTO` is niclass if we want a bit more perf, but I don't expect it to be noticeable.

**Comment by DarkDimius** [comment]
How is this different from https://github.com/sorbet/sorbet/pull/1757? Why has the other failed while this succeeded?

**Comment by elliottt** [comment]
Good catch, they're the same. I had lost track of what I had already submitted with all the chaos of travel.

My suspicion is that the change that makes this work is #1756, so I'm going to close #1757 and dig into that a bit.

**Comment by elliottt** [comment]
Actually, on closer inspection the two differ in a pretty fundamental way: #1757 compares the temporary against thisTO.type, while this branch compares against otherTO.type:

https://github.com/sorbet/sorbet/pull/1757/files#diff-ff7fc5031249f34da61e4335f7fbab16R595

vs:

https://github.com/sorbet/sorbet/pull/1772/files#diff-ff7fc5031249f34da61e4335f7fbab16R594

I remember you saying that `Types::any` was left-biased, so it's surprising that this version works while the other raises errors.

**Comment by DarkDimius** [comment]
Let's make it agnostic of biasing by doing two reference comparisons? `temp == otherTO.type || temp == thisTO.type;;

File: core/types/types.cc [hunk 2]
```diff
@@ -214,12 +214,23 @@ TypePtr Types::dropSubtypesOf(Context ctx, const TypePtr &from, SymbolRef klass)
 }
 
 bool Types::canBeTruthy(Context ctx, const TypePtr &what) {
-    if (what->isUntyped()) {
-        return true;
-    }
-    auto truthyPart =
-        Types::dropSubtypesOf(ctx, Types::dropSubtypesOf(ctx, what, Symbols::NilClass()), Symbols::FalseClass());
-    return !truthyPart->isBottom(); // check if truthyPart is empty
+    bool isTruthy = false;
```
Comment (DarkDimius) [inline core/types/types.cc]: can we have a conservative default?

File: core/types/types.cc [hunk 3]
```diff
@@ -214,12 +214,23 @@ TypePtr Types::dropSubtypesOf(Context ctx, const TypePtr &from, SymbolRef klass)
...
+            auto sym = c->symbol;
+            isTruthy = sym == core::Symbols::untyped() ||
+                       (sym != core::Symbols::FalseClass() && sym != core::Symbols::NilClass());
+        },
+        [&](AppliedType *c) {
+            auto sym = c->klass;
+            isTruthy = sym == core::Symbols::untyped() ||
+                       (sym != core::Symbols::FalseClass() && sym != core::Symbols::NilClass());
+        },
+        [&](ProxyType *c) { isTruthy = canBeTruthy(ctx, c->underlying()); }, [&](Type *) {});
```
Comment (DarkDimius) [inline core/types/types.cc]: ditto, let's make the `Type *` conservative.

**Pull request merged**